ssl/ssl_lib.c

```diff
@@ -2205,20 +2212,41 @@ long SSL_CTX_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
     case SSL_CTRL_SET_SESS_CACHE_SIZE:
         if (larg < 0)
             return 0;
-        l = (long)ctx->session_cache_size;
-        ctx->session_cache_size = (size_t)larg;
+        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+        l = (long)ctx->session_cache->size;
+        ctx->session_cache->size = (size_t)larg;
+        CRYPTO_THREAD_unlock(ctx->session_cache->lock);
+        CRYPTO_THREAD_unlock(ctx->lock);
         return l;
     case SSL_CTRL_GET_SESS_CACHE_SIZE:
-        return (long)(ctx->session_cache_size);
+        CRYPTO_THREAD_read_lock(ctx->lock);
+        l = (long)ctx->session_cache->size;
+        CRYPTO_THREAD_unlock(ctx->lock);
+        return l;
+
     case SSL_CTRL_SET_SESS_CACHE_MODE:
-        l = ctx->session_cache_mode;
-        ctx->session_cache_mode = larg;
-        return (l);
+        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+        l = ctx->session_cache->mode;
+        ctx->session_cache->mode = larg;
```

Comment:
Of course existing behavior should not change when no shared session cache is used.
I just wonder if it should be allowed that changing a property on one ctx
changes all other ctx using the same session cache.
For that purpose it would be more appropriate to set the property on the cache itself.